### PR TITLE
[ROCm] Enable FP8 KV-cache and relax constraints for RDNA4 custom paged attention

### PR DESCRIPTION
**Pull request by laudney**
## Summary

Enable FP8 KV-cache support in the gfx12 (RDNA4) custom ROCm
paged-attention kernel. The kernel dequantizes FP8 K and V data in
software, applies the K and V scales, and fixes the QK WMMA lane layout.

FP8 KV cache uses half the storage of a 16-bit KV cache. On a Radeon AI
PRO R9700, prior model validation increased the available KV-cache capacity
from about 132K to 264K tokens for Qwen3-Coder-30B AWQ.

## Changes

- Convert FP8 e4m3 K and V vectors to FP16 or BF16 before WMMA.
- Keep query loading independent of the KV-cache element size.
- Correct the FP8 QK row and lane alignment with a cross-row shuffle.
- Apply `k_scale` to QK logits and `v_scale` to the attention output.
- Allow FP8 KV cache, block size 32, and GQA ratios 1-2 only on
  `_ON_GFX12X`; keep the existing gfx11 limits.
- Keep head size 128 and no ALiBi for the Navi custom kernel.
- Pass the optional FP8 output scale through the Navi launcher.

## Duplicate-work check

Open PRs #46191 and #46192 were checked. #46191 changes the Triton
chunked-prefill fallback. #46192 changes AITER unified attention. Neither PR
implements FP8 KV cache in this custom HIP paged-attention kernel.

## Validation

- Rebased onto upstream `main` at `a3b86752f`.
- All pre-commit hooks passed for the three changed files.
- The editable ROCm native extension built from commit `602e34c5a1`.
- `test_paged_attention` ran on one Radeon AI PRO R9700 (`gfx1201`) through
  the repository virtual environment with AMDSMI initialized before pytest:
  **12 passed, 36 skipped in 27.50 seconds**. The skips are the unsupported
  head-size and ALiBi combinations.

Prior independent R9700 validation on ROCm 7.2.3 reported 136 passed and 56
skipped, and TPOT improved in all 10 tested Qwen3-30B-A3B workloads. The mean
TPOT reduction was 32.81%. Prior model checks also covered
Qwen3-Coder-30B-A3B AWQ, Qwen3-14B-FP8, and GPT-OSS-20B MXFP4 with FP8 KV
cache and correct inference output.

No new end-to-end model evaluation was run after this rebase. The native
kernel test covers the rebased code on the target hardware, and the earlier
model and performance results remain recorded in the PR discussion.

## AI assistance

AI assistance was used to analyze duplicate work, rebase the branch, resolve
the test conflict, review the updated diff, and run validation. The human
submitter must review every changed line and be able to explain and defend the
change end to end.

### CODE REVIEW

File: vllm/platforms/rocm.py [hunk 1]
```diff
@@ -178,16 +179,19 @@ def use_rocm_custom_paged_attention(
...
+        block_size_ok = block_size == 16 or (_ON_GFX12 and block_size == 32)
         return (
             _ON_GFX1X
             and (sliding_window == 0 or sliding_window == (-1, -1))
             and (qtype == torch.half or qtype == torch.bfloat16)
             and head_size == 128
-            and block_size == 16
-            and (gqa_ratio >= 3 and gqa_ratio <= 16)
+            and block_size_ok
+            and (gqa_ratio >= 1 and gqa_ratio <= 16)
```
Comment (tjtanaa) [inline vllm/platforms/rocm.py]: won't this affect the condition of `gfx11x`?

Comment (tjtanaa) [reply vllm/platforms/rocm.py]: The custom paged attention kernels implementation for gfx12x and gfx11x are different.

File: tests/kernels/attention/test_attention.py [hunk 2]
```diff
@@ -139,14 +139,13 @@ def test_paged_attention(
...
-        and (
-            kv_cache_dtype == "fp8" or head_size != 128 or block_size != 16 or use_alibi
+    if version == "rocm" and current_platform.is_navi():
+        # gfx12 (RDNA4) supports FP8 KV cache via software dequant
+        fp8_unsupported = (
+            kv_cache_dtype == "fp8" and not current_platform.supports_fp8()
         )
-    ):
-        pytest.skip()
+        if fp8_unsupported or head_size != 128 or block_size != 16 or use_alibi:
```
Comment (tjtanaa) [inline tests/kernels/attention/test_attention.py]: I thought on `gfx12` it supports block size of `32` as stated in

`block_size_ok = block_size == 16 or (_ON_GFX12 and block_size == 32)`

**Comment by laudney** [comment]
Rebased onto latest main. Two small conflicts in `vllm/platforms/rocm.py`:

- Upstream renamed the gfx12 sentinel to `_ON_GFX12X` (matching the `_ON_GFX9`/`_ON_GFX1X` naming convention); folded our `_ON_GFX12` definition into the upstream one and updated all references in this PR accordingly.
- Merged with the follow-up `gqa_min = 1 if _ON_GFX12X else 3` from commit d4f6c0d69 so the gfx11 retains `>= 3` behavior while gfx12 keeps the `>= 1` relaxation.

Pre-commit clean. Ready for re-review @tjtanaa.

**Comment by laudney** [comment]
@tjtanaa Rebased onto current main and squashed the 3 iteration commits into one (`a54231f`). All three of your inline comments are addressed in this state: gfx11 retains `gqa_ratio >= 3` / `block_size == 16`, gfx12 gets the relaxations, and the test gate now allows `block_size == 32` only when `is_gfx12`. The rebase also absorbed upstream's `__HIP__GFX12__` → `__GFX12__` macro rename in `csrc/rocm/attention.cu`. Could you re-review? @mgoin / @gshtras — could one of you apply `ready` so the first-contributor CI gate clears?

**Comment by keneoneth** [comment]
I tested the changes from this PR after rebasing them onto
[`bdbf08fc0277c63b90478fae2566aaacdaa71d9b`](https://github.com/vllm-project/vllm/commit/bdbf08fc0277c63b90478fae2566aaacdaa71d9b). From the analysis results below, this PR is important for RDNA4 users because it enables the optimized ROCm paged-attention path with FP8 KV cache, providing substantial decode-performance gains.


### Environment

- GPU: AMD Radeon AI PRO R9700 (`gfx1201`)
- ROCm: 7.2.3
- Base commit: `bdbf08fc0277c63b90478fae2566aaacdaa71d9b`
- Model: [QW](https://huggingface.co/Qwen/Qwen3-30B-A3B-Instruct-2507)
- Tensor parallel size: 4
- KV-cache dtype: `fp8`

### Correctness Validation

I ran the paged-attention unit-test matrix locally on AMD Radeon AI PRO R9700 (gfx1201).

pytest -v vllm/tests/kernels/attention/test_attention.py::test_paged_attention

Result: 136 tests passed, 56 tests skipped
Note: 56 tests are skipped due to head size not aligned to 128 requirement or ALiBi feature not being supported on Navi platform

### Performance Comparison

| Scenario | Workload (input/output tokens) | Base TTFT (ms) | PR TTFT (ms) | TTFT change | Base TPOT (ms) | PR TPOT (ms) | TPOT change |
| :--- | :--- | :---: | :---: | :---: | :---: | :---: | :---: |
| Multi-batch | 512/128 | 134.88 | 133.59 | −0.96% | 76.27 | 44.69 | −41.41% |
| Multi-batch | 1024/256 | 793.63 | 182.35 | −77.02% | 49.47 | 44.70 | −9.64% |
| Multi-batch | 4096/512 | 515.09 | 404.70 | −21.43% | 50.02 | 27.35 | −45.32% |
| Multi-batch | 8192/1024 | 1440.68 | 1167.60 | −18.95% | 78.78 | 31.76 | −59.69% |
| Multi-batch | 2048/128 | 177.39 | 165.65 | −6.62% | 63.19 | 58.19 | −7.91% |
| Single request | 512/128 | 48.27 | 49.72 | +3.00% | 24.34 | 22.58 | −7.23% |
| Single request | 1024/256 | 60.67 | 61.24 | +0.94% | 27.03 | 22.60 | −16.39% |
| Single request | 4096/512 | 123.15 | 123.00 | −0.12% | 44.04 | 22.72 | −48.41% |
| Single request | 8192/1024 | 208.50 | 207.00 | −0.72% | 63.98 | 23.05 | −63.97% |
| Single request | 2048/128 | 59.57 | 60.54 | +1.63% | 31.50 | 22.63 | −28.16% |

### Summary
| Metric | Average latency change vs base commit |
| :--- | :---: |
| Mean TTFT | −12.03% |
| Mean TPOT | −32.81% |

Across the 10 tested workloads, TPOT improved in every case. Multi-batch TTFT also improved in all five workloads. The improvement is especially significant for long-context RAG workloads. Compared with the base commit, TPOT decreased from 78.78 ms to 31.76 ms in the multi-batch case and from 63.98 ms to 23.05 ms in the single-request case. This corresponds to approximately 2.48× and 2.78× higher output token throughput, respectively. Overall, the results demonstrate promising performance gains.

**Comment by laudney** [comment]
@tjtanaa @dllehr-amd — rebased onto latest `main` and now mergeable. This enables FP8 KV-cache on the RDNA4 (gfx12) custom paged-attention kernel (≈2× KV-cache capacity → longer context / more concurrent sequences) and fixes a real FP8 QK WMMA16x16x16 head-element misalignment that produced degenerate decode output; both verified on gfx1201 hardware. Would appreciate a review when you have a chance.

**Comment by laudney** [comment]
Rebased onto current upstream main at a3b86752f and force-pushed as one commit, 602e34c5a1. The ROCm native extension built successfully, and the complete local test_paged_attention matrix on an R9700 (gfx1201) finished with 12 passed and 36 deliberate skips. I also checked #46191 and #46192; neither covers this custom HIP FP8 KV-cache path. The pre-run gate stops CI because this account has 2 merged PRs and the PR has no ready or verified label. @tjtanaa @dllehr-amd @mgoin, could you re-review and apply ready, ready-run-all-tests, or verified if appropriate?

**Comment by cadamcat** [comment]
@laudney @tjtanaa The gfx11 scoping question in this PR is one I have hardware and measurements for, so here they are.

The review concern that led to `gqa_min = 1 if _ON_GFX12X else 3` was:

> won't this affect the condition of `gfx11x`?
>
> The custom paged attention kernels implementation for gfx12x and gfx11x are
> different.

Both points are right, and the second is why this comment is only about gfx11. I measured the gfx11 implementation on 2x RX 7900 XT (gfx1100). Nothing here says anything about gfx12.

**On gfx11, at the excluded ratios, the CK kernel is both correct and faster.**

Correctness first, since that is what "the implementations are different" makes worth checking. Scored against an fp32 reference computed from the pre-paging KV, relative L-inf across 1K to 32K: CK sits at 2.9e-3 to 4.7e-3 and the Triton fallback at 2.1e-3 to 3.4e-3. Same band, no drift with context. An A100 running the byte-identical Triton source lands in the same band, so that is bf16's floor for a sum this long rather than anything architectural.

Speed, Triton time over CK time, so above 1.0 means the currently-excluded kernel is the faster one:

| shape | gqa | gated in today | 1K | 4K | 16K | 32K |
|---|---:|---|---:|---:|---:|---:|
| 8/8 | 1 | no | 2.06x | 2.87x | 4.11x | 4.83x |
| 8/4 | 2 | no | 2.31x | 3.21x | 6.53x | 7.28x |
| 32/16 | 2 | no | 2.14x | 1.84x | 2.71x | 2.86x |
| 12/4 | 3 | yes | 2.40x | 3.20x | 6.36x | 7.23x |
| 16/4 | 4 | yes | 2.40x | 3.20x | 6.42x | 7.40x |

The gated-in rows are the control. gqa 3 and 4 show the advantage the bound was written to capture, and gqa 1 and 2 show the same advantage at the same magnitude being declined. For context on where the 3 came from, [#17004](https://github.com/vllm-project/vllm/pull/17004) states it as a performance observation rather than a capability limit:

> It supports gqa_ratio up to 16, and shows performance gains over the existing
> kernel when gqa_ratio is 3 or higher. Therefore, it is enabled for gqa_ratio
> values between 3 and 16.

**End to end**, `gemma-3-27b-it-w4a16` (32/16 heads, head_dim 128, so gqa 2),
TP=2, CUDA graphs on, decode tok/s by 64-vs-8 differencing, one fresh container per cell, `gqa_min` set to 1 at source:

| ctx | stock | gqa_min=1 | |
|---|---:|---:|---:|
| 1024 | 41.44 | 42.57 | 1.027x |
| 8192 | 21.36 | 23.88 | 1.118x |
| 32768 | 8.00 | 9.55 | **1.194x** |

The curve is structural, not noise. gemma-3-27b has `sliding_window_pattern: 6`, so only the ~10 full-attention layers of 62 can move; the ~52 sliding layers fail `sliding_window == 0` in both arms. At 1K those dominate; at 32K a full layer scans the whole context while a sliding layer scans 1024, so the layers that moved carry most of the decode KV traffic.

Routing was recorded from inside the TP workers rather than inferred, in every cell and on both ranks:

| layer | stock | gqa_min=1 |
|---|---|---|
| full attention, `window=0` | `use_custom=False` | `use_custom=True` |
| sliding, `window=1023` | `use_custom=False` | `use_custom=False` |

Those same records show the full-attention layers arriving with head_size 128, block_size 16, native KV layout and a power-of-two block, so on this model `gqa_ratio` is the only condition they fail.

**Two things to weigh against it.** vllm#53856 is fixing an unmasked V-cache
padding read in this kernel, where a NaN in a slot past `seq_len` poisons the output; I reproduced that on gfx1100 at a gated-in ratio and reported it there. Lowering `gqa_min` widens the set of shapes reaching the kernel, so it widens exposure to that until #53856 lands. It does not introduce it, since every gfx11 model with gqa_ratio 3 to 16 is already on this path. And the measurements are on vLLM 0.23.1.dev1 on gfx1100, not on this branch and not on current main.

If this is useful and you would rather not carry it yourself, would you be open to a small PR against `mmonad/vllm:feat/rocm-rdna4-custom-attn-fp8kv` changing `gqa_min` to cover gfx11? Either way works. I did not want to push at your branch uninvited, given the gfx11 scoping was a deliberate review outcome rather than an oversight.

Probes, every measured cell, and the worker-side routing records: https://github.com/cadamcat/dual-radeon-vllm/tree/main/benchmarks/vllm-50603

*Edited 2026-08-28. The question this comment ends on is answered from my side rather than yours: I filed the gfx11 widening as [#54210](https://github.com/vllm-project/vllm/pull/54210) against `main`, one line, because the bound is not gfx12-specific and neither is the evidence for widening it. It touches the same expression this PR rewrites — once `gqa_min = 1 if _ON_GFX12X else 3` is in, widening gfx11 is `else 3` to `else 1`, so whichever of the two rebases second should expect that hunk rather than meet it as a conflict. If you would rather carry the gfx11 widening here, #54210 can be closed in favour of it.*

*Two figures above have also been superseded. #54210 carries the same sweep re-run on 0.27, 30 cells twice: no cell of the sixty has CK slower, and the excluded ratios give 2.03-4.09x at gqa 1 and 1.70-6.05x at gqa 2, against 2.20-6.05x and 2.35-6.04x for the admitted 3 and 4. End to end it is 1.024x / 1.084x / 1.167x at 1K / 8K / 32K, against the single-pass 1.027 / 1.118 / 1.194 in the table above; a single pass was not enough for that arm. The 0.23.1 rows here are left as measured.*
